File: src/main/java/cspom/compiler/PredicateScanner.java
package cspom.compiler;

import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;
import java.util.regex.Pattern;

public final class PredicateScanner {

	public static final Pattern INTEGER = Pattern.compile("[0-9]*");

	public static final Pattern IDENTIFIER = Pattern.compile("[a-zA-Z_]\\w*");

	private PredicateScanner() {

	}

	public static Node scan(String expression) throws PredicateParseException {
		final StringTokenizer st = new StringTokenizer(expression, " (),", true);
		final Deque<Node> stack = new LinkedList<Node>();
		Node currentNode = new Node();
		while (st.hasMoreElements()) {
			final String token = st.nextToken();
			if (" ".equals(token)) {
				continue;
			} else if ("(".equals(token)) {
				if (currentNode.operator == null) {
					throw new PredicateParseException("Empty operator");
				}
				final Node newNode = new Node();
				currentNode.child = newNode;
				stack.push(currentNode);
				currentNode = newNode;
			} else if (")".equals(token)) {
				if (stack.isEmpty()) {
					throw new PredicateParseException("Too many )s");
				}
				currentNode = stack.pop();

			} else if (",".equals(token)) {
				if (currentNode.operator == null) {
					throw new PredicateParseException("Empty argument");
				}
				final Node newNode = new Node();
				currentNode.sibling = newNode;
				currentNode = newNode;
			} else {
				if (currentNode.operator != null) {
					throw new PredicateParseException("Delimiter expected in "
							+ currentNode + " (" + expression + ")");
				}
				currentNode.operator = token;
			}
		}
		return currentNode;
	}

	public static final class Node {
		private Node sibling;
		private Node child;
		private String operator;

		public Node getSibling() {
			return sibling;
		}

		public Node getChild() {
			return child;
		}

		public String getOperator() {
			return operator;
		}

		public String toString() {
			final StringBuilder stb = new StringBuilder();
			tree(stb, 0);
			return stb.toString();
		}

		public boolean isInteger() {
			if (operator == null) {
				return false;
			}
			return INTEGER.matcher(operator).matches();
		}

		public boolean isIdentifier() {
			if (operator == null) {
				return false;
			}
			return IDENTIFIER.matcher(operator).matches();
		}

		public boolean isLeaf() throws PredicateParseException {
			if (child == null) {
				if (!isInteger() && !isIdentifier()) {
					throw new PredicateParseException(
							"Leaves should be variables or constants");
				}
				return true;
			}
			return false;
		}

		private void tree(StringBuilder stb, int level) {
			for (int i = level; --i >= 0;) {
				stb.append("-");
			}
			stb.append(operator).append("\n");
			if (child != null) {
				child.tree(stb, level + 1);
			}
			if (sibling != null) {
				sibling.tree(stb, level);
			}
		}
	}
}

File: src/main/scala/cspom/compiler/PredicateScanner.scala
package cspom.compiler;

import scala.collection.mutable.Stack
import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;
import java.util.regex.Pattern;

final object PredicateScanner {

  val INTEGER = Pattern.compile("-?[0-9]*");

  val IDENTIFIER = Pattern.compile("[a-zA-Z_]\\w*");

  def scan(expression: String): PredicateNode = {
    val st = new StringTokenizer(expression, " {}(),", true);
    val stack = new Stack[PredicateNode];
    var currentNode = new PredicateNode;
    var parameters: StringBuilder = null;
    
    while (st.hasMoreElements()) {
      val token = st.nextToken();
      if ("}" == token) {
        currentNode.parameters = Some(parameters.toString)
        parameters = null;
      } else if (parameters != null) {
        parameters.append(token);
      } else token match {
        case " " =>
        case "{" =>
          assume(currentNode.operator.isDefined, "Empty operator");
          parameters = new StringBuilder();
        case "(" =>
          assume(currentNode.operator.isDefined, "Empty operator");
          val newPredicateNode = new PredicateNode();
          currentNode.child = Some(newPredicateNode)
          stack.push(currentNode);
          currentNode = newPredicateNode;
        case ")" =>
          assume(!stack.isEmpty, "Too many )s");
          currentNode = stack.pop();
        case "," =>
          assume(currentNode.operator.isDefined, "Empty argument");
          val newPredicateNode = new PredicateNode();
          currentNode.sibling = Some(newPredicateNode)
          currentNode = newPredicateNode;
        case _ =>
          assume(currentNode.operator.isEmpty, "Delimiter expected in "
            + currentNode + " (" + expression + ")");
          currentNode.operator = Some(token)
      }
    }
    currentNode;
  }

}
